src/v/cluster/rm_stm.cc

```diff
@@ -1327,306 +1328,115 @@ ss::future<result<kafka_result>> rm_stm::transactional_replicate(
+        if (result.error() == errc::sequence_out_of_order) {
+            auto barrier = co_await _raft->linearizable_barrier();
+            if (!barrier) {
                 co_return errc::not_leader;
             }
         }
-    } else {
-        if (!is_sequence(session->tail_seq, bid.first_seq)) {
-            auto barrier = co_await _raft->linearizable_barrier();
-            if (barrier) {
-                // there is a gap between the request's seq number
-                // and the seq number of the latest inflight request
-                // may happen when:
-                //   - batches were reordered on the wire
-                //   - client is misbehaving
-                vlog(
-                  _ctx_log.warn,
-                  "[pid: {}] sequence number gap detected. batch first "
-                  "sequence: "
-                  "{}, last inflight sequence: {}",
-                  bid.pid,
-                  bid.first_seq,
-                  session->tail_seq);
-                co_return errc::sequence_out_of_order;
-            }
-            co_return errc::not_leader;
+        if (
+          _raft->is_leader() && _raft->term() == synced_term
+          && (result.error() == errc::replication_error || result.error() == errc::timeout)) {
+            co_await _raft->step_down(
+              "Failed replication during idempotent replicate.");
         }
+        co_return result.error();
     }
+    co_return result.value();
+}
 
-    // updating last observed seq, since we hold the mutex
-    // it's guranteed to be latest
-    session->tail_seq = bid.last_seq;
-
-    auto request = ss::make_lw_shared<inflight_request>();
-    request->last_seq = bid.last_seq;
-    request->is_processing = true;
-    session->cache.push_back(request);
-
-    // request comes in the right order, it's ok to replicate
-    std::optional<raft::replicate_stages> ss;
-    try {
-        ss = _raft->replicate_in_stages(synced_term, std::move(br), opts);
-        co_await std::move(ss->request_enqueued);
-    } catch (...) {
+ss::future<result<kafka_result>> rm_stm::do_idempotent_replicate(
+  model::term_id synced_term,
+  producer_ptr producer,
+  model::batch_identity bid,
+  model::record_batch_reader br,
+  raft::replicate_options opts,
+  ss::lw_shared_ptr<available_promise<>> enqueued,
+  ssx::semaphore_units units) {
+    using ret_t = result<kafka_result>;
+    auto request = producer->try_emplace_request(bid, synced_term);
+    if (!request) {
+        co_return request.error();
+    }
+    auto req_ptr = request.value();
+    if (req_ptr->state() != request_state::initialized) {
+        // request already in progress/ completed.
+        co_return co_await req_ptr->result();
+    }
+
+    req_ptr->mark_request_in_progress();
+    auto stages = _raft->replicate_in_stages(synced_term, std::move(br), opts);
+    auto req_enqueued = co_await ss::coroutine::as_future(
+      std::move(stages.request_enqueued));
+    units.return_all();
```

Comment:
the order in which we released the lock and step down used to matter, see code around `// we may not care about the requests waiting on the lock`; are you sure it doesn't matter now?